src/basic/fs-util.c

```diff
@@ -1135,3 +1140,53 @@ int xopenat(int dir_fd, const char *path, int flags, mode_t mode) {
 
         return TAKE_FD(fd);
 }
+
+int xopenat_lock(int dir_fd, const char *path, int flags, mode_t mode, LockType locktype, int operation) {
+        _cleanup_close_ int fd = -EBADF;
+        int r;
+
+        assert(dir_fd >= 0 || dir_fd == AT_FDCWD);
+        assert(path);
+        assert(IN_SET(operation & ~LOCK_NB, LOCK_EX, LOCK_SH));
+
+        /* POSIX/UNPOSIX locks don't work on directories. */
+        if (flags & O_DIRECTORY)
+                assert(locktype == LOCK_BSD);
```

Comment:
i mean, if i get this right: if the inode already exists and is a directory, and you do *not* specify O_DIRECTORY, then the xopenat() below will succeed, but then the actual lock fail. I think it would be wise to return the same error here early if we see O_DIRECTORY, as in that case, so that we have somewhat systematic behaviour